test/mpi/util/mtest.c

```diff
@@ -227,6 +227,18 @@ void MTest_Finalize(int errs)
     MPI_Finalize();
 
     MTest_finalize_thread_pkg();
+#ifdef HAVE_ZE
```

Comment:
Can we turn this block into a utility function? Say `MTest_finalize_gpu`? I think it should be possible to avoid exposing all the ZE-related variables, if so.

Reply:
I can do that.

Reply:
I've added `MTest_finalize_gpu` function in `mtest.c` and `mtest.cxx`. I cannot put this implementation in `mtest_common.c`, since several tests are only linked with `mtest.c` and `mtest.cxx` objects.

Reply:
I don't understand how that is possible. The CUDA tests did not require this.

Reply:
The issue is here from my local test:
```
make[3]: Entering directory `/home/caochong/build/mpich-ch4-build/test/mpi/f77/ext'
/bin/sh ../../libtool  --tag=CC   --mode=link /home/caochong/opt/mpich-ch4/bin/mpicc     -o c2fmult c2fmult.o ../../util/mtest.o ../../util/mtest_single.o -lcudart  -L/usr/local/cuda/lib64 -L/usr/local/cuda/lib
libtool: link: /home/caochong/opt/mpich-ch4/bin/mpicc -o c2fmult c2fmult.o ../../util/mtest.o ../../util/mtest_single.o  -lcudart -L/usr/local/cuda/lib64 -L/usr/local/cuda/lib
../../util/mtest.o: In function `MTest_Finalize':
/home/caochong/git/mpich-ofi/test/mpi/util/mtest.c:230: undefined reference to `MTest_finalize_gpu'
collect2: error: ld returned 1 exit status
```

You can see `c2fmult` is not linked with `mtest_common.o`, so the `MTest_finalize_gpu()` should be implemented inside `mtest.c`.

Even if we think CUDA tests actually don't need this `finalize_gpu()` function, it will become the same issue for ZE.

Reply:
Let me see if I can add `mtest_common.o` into dependencies of `c2fmul`, that might be the better way to solve it.

Reply:
I guess this really boils down to the fact that we link object files directly instead of building an mtest.la libtool library. Since we relented and required libtool support with the integration of dtpools, it's probably time someone went and cleaned up the mtest utilities. @shawnccx I suppose this isn't your problem to solve. If you can make the example work, great. If not, I'll re-review and merge it as it is.

Reply:
> Let me see if I can add `mtest_common.o` into dependencies of `c2fmul`, that might be the better way to solve it.

You  should be able to with something  like `c2fmul_LDADD`. But it'd be nice to reduce the number of tests require custom rules.

@raffenet Do you recall the difference between `mtest.c` and `mtest_common.c`? Intuitively,  if something is named  as `_common`, it should always be linked in.

Reply:
> I guess this really boils down to the fact that we link object files directly instead of building an mtest.la libtool library.

Now we keep adding objects, I agree it it a good idea to make use of at least convenience library. We probably need multiple of them, such as `_thread.la`, `_f77.la` etc.

Reply:
I think their existence predates my joining the project. My guess is that `mtest_common.c` was intended for utilities that could be used in any language.